src/stt/transform_speech_data.py

```diff
@@ -0,0 +1,370 @@
+import re
+import shutil
+import sys
+from typing import Optional, Union
+
+import click
+import pandas as pd
+from constants import (
+    CHARS_TO_IGNORE_REGEX,
+    NUM_PROC,
+    PROCESSED_DIR,
+    RAW_DATA_DIR,
+    SAMPLING_RATE,
+    SRC_DIR,
+    VOCAB_DIR,
+)
+from datasets import Audio, Dataset
+from datasets.arrow_dataset import Batch, Example
+from transformers import Wav2Vec2Processor
+from unidecode import unidecode
+from utils import load_processor
+
+sys.path.insert(0, SRC_DIR)
+from decorators import log_function_name  # noqa: E402
+
+
+def remove_special_characters(
+    batch: Example, train: Optional[bool] = True
+) -> Example:
+    """
+    Remove special characters from the dataset
+
+    Parameters
+    ----------
+    batch : Example
+        Batch of data
+    train : Optional[bool], optional
+        Checks if batch belongs to train dataset, by default True
+
+    Returns
+    -------
+    Example
+        Batch with special characters removed
+    """
+    batch["sentence"] = (
+        re.sub(
+            CHARS_TO_IGNORE_REGEX, "", unidecode(batch["sentence"])
+        )
+        .lower()
+        .strip()
+    )
+    if train:
+        batch["sentence"] += " "
+
+    return batch
+
+
+def transform_dataset(
+    batch: Example, processor: Wav2Vec2Processor
+) -> Example:
+    """
+    Transform dataset to be used by the model
+
+    Parameters
+    ----------
+    batch : Example
+        Batch of data
+    processor : Wav2Vec2Processor
+        Processor to use
+
+    Returns
+    -------
+    Example
+        Batch with transformed data
+    """
+    audio = batch["audio"]
+    batch["input_values"] = processor(
+        audio["array"], sampling_rate=audio["sampling_rate"]
+    ).input_values[0]
+    batch["labels"] = processor(text=batch["sentence"]).input_ids
+    return batch
+
+
+@log_function_name
+def create_vocab(
+    folder_path: str,
+    train_ds: Dataset,
+    val_ds: Dataset,
+    test_ds: Dataset,
+) -> None:
+    """
+    Create vocab file from datasets
+
+    Parameters
+    ----------
+    folder_path : str
+        Path to the folder where the vocab file will be saved
+    train_ds : Dataset
+        Train dataset
+    val_ds : Dataset
+        Validation dataset
+    test_ds : Dataset
+        Test dataset
+    """
+
+    def extract_all_chars(
+        batch: Batch,
+    ) -> dict[str, list[Union[list[str], str]]]:
+        """
+        Extract all characters from a batch
+
+        Parameters
+        ----------
+        batch : Batch
+            Batch to extract characters from
+
+        Returns
+        -------
+        dict[str, list[Union[list[str], str]]]
+            Dictionary containing all characters from the batch
+        """
+        all_text = " ".join(batch["sentence"])
+        vocab = list(set(all_text))
+        return {"vocab": [vocab], "all_text": [all_text]}
+
+    vocab_train = train_ds.map(
+        extract_all_chars,
+        batched=True,
+        batch_size=-1,
```

Comment:
batched true to size 1 - macht das sinn? oder kann man dann direkt batch false sagen

Reply:
Das gibt quasi die bestmögliche batch size an